File: dictionaries-and-hashmaps/00-RansomNote.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the checkMagazine function below.
def checkMagazine(magazine, note):
    availableWords = {}
    for m in magazine:
        if m in availableWords:
            availableWords[m] += 1
        else:
            availableWords[m] = 1
            
    
    result = "Yes"
    for n in note:
        if n not in availableWords:
            result = "No"
            break
        if availableWords[n] > 0:
            availableWords[n] -= 1
        else:
            result = "No"
            break
            
    return result
            
        
        

if __name__ == '__main__':
    mn = input().split()

    m = int(mn[0])

    n = int(mn[1])

    magazine = input().rstrip().split()

    note = input().rstrip().split()

    result = checkMagazine(magazine, note)
    
    fptr = open(os.environ['OUTPUT_PATH'], 'w')
    
    fptr.write(str(result) + '\n')

    fptr.close()

File: dictionaries-and-hashmaps/00-RansomNote.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the checkMagazine function below.
    static void checkMagazine(String[] magazine, String[] note) {
        Hashtable<String, Integer> wordDict = new Hashtable<String, Integer>();
        boolean canForm = true;

        // Fill dictionary
        for (String s: magazine) {
            Integer word = wordDict.get(s);
            if (word == null) {
                wordDict.put(s, 1);
            } else {
                wordDict.put(s, wordDict.get(s) + 1);
            }
        }

        for (String n: note) {
            if (wordDict.containsKey(n) && wordDict.get(n) > 0) {
                wordDict.put(n, wordDict.get(n) - 1);
            } else {
                canForm = false;
                break;
            }
        }

        System.out.println(canForm ? "Yes" : "No");
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        String[] mn = scanner.nextLine().split(" ");

        int m = Integer.parseInt(mn[0]);

        int n = Integer.parseInt(mn[1]);

        String[] magazine = new String[m];

        String[] magazineItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < m; i++) {
            String magazineItem = magazineItems[i];
            magazine[i] = magazineItem;
        }

        String[] note = new String[n];

        String[] noteItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < n; i++) {
            String noteItem = noteItems[i];
            note[i] = noteItem;
        }

        checkMagazine(magazine, note);

        scanner.close();
    }
}
